crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
   }
+
   // Caller is getting parameter values.
   if (sig == NULL) {
-    *siglen = DILITHIUM3_SIGNATURE_BYTES;
+    *siglen = nistdsa->signature_len;
     return 1;
   }
 
-  if (*siglen < DILITHIUM3_SIGNATURE_BYTES) {
+  if (*siglen < nistdsa->signature_len) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
     return 0;
   }
 
-  if (ml_dsa_65_sign(sig, siglen, tbs, tbslen, NULL, 0, key->priv) != 0) {
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+      ctx->pkey->pkey.nistdsa_key == NULL ||
+      ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+      }
+
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
+  if (!key->secret_key) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (nistdsa->method->sign(sig, siglen, tbs, tbslen, NULL,
+                            0, key->secret_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, ERR_R_INTERNAL_ERROR);
     return 0;
   }
   return 1;
 }
 
-static int pkey_dilithium3_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
+static int pkey_nistdsa_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
                                           size_t siglen, const uint8_t *tbs,
                                           size_t tbslen) {
-  if (ctx == NULL || ctx->pkey->pkey.ptr == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
-    return 0;
+  NISTDSA_PKEY_CTX *dctx = ctx->data;
+  const NISTDSA *nistdsa = dctx->nistdsa;
+
+  if (nistdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+
+    nistdsa = NISTDSA_KEY_get0_sig(ctx->pkey->pkey.nistdsa_key);
   }
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+    ctx->pkey->pkey.nistdsa_key == NULL ||
+    ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+    }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
 
-  if (siglen != DILITHIUM3_SIGNATURE_BYTES ||
-      ml_dsa_65_verify(tbs, tbslen, sig, siglen, NULL, 0, key->pub) != 0) {
+  if (siglen != nistdsa->signature_len ||
+      nistdsa->method->verify(tbs, tbslen, sig, siglen,
+                              NULL, 0, key->public_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
+    return 0;
+      }
+
+  return 1;
+}
+
+// This function sets nistdsa parameters defined by |nid| in |pkey|.
+// If |pkey| already has a public key set, this public key is preserved.
+int EVP_PKEY_nistdsa_set_params(EVP_PKEY *pkey, int nid) {
+  const NISTDSA *nistdsa = SIG_find_dsa_by_nid(nid);
+
+  if (nistdsa == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+
+  // if the public key has already been set either by EVP_parse_public_key or
```

Comment:
this doesn't sound safe. For example, what happens to the security of ML-DSA if you use one variant parameters with another variant keys?

Reply:
Seconding that, we can check whether the params of the existing key match the params being set. But that seems not needed and it should be an error as is done in `EVP_PKEY_CTX_nistdsa_set_params`. What is this function used for? Does it need to be exported? `EVP_PKEY_kem_set_params` in https://github.com/aws/aws-lc/blob/main/crypto/fipsmodule/evp/p_kem.c#L357 is static and is used in importing keys.

Reply:
Okay, I moved `EVP_PKEY_pqdsa_set_params` from evp.h so it is no longer exported (I was having problems getting the benchmarking suite speed.cc to find the new API and I thought this was because it needed to be exported, but I just needed to increment the API level) in [e47ef83](https://github.com/aws/aws-lc/pull/1963/commits/e47ef838c72641966a62cf59044e44fcae7f5009).

As to what this function is used for, and how that differs to the KEM param set function:

Both `EVP_PKEY_pqdsa_set_params` and `EVP_PKEY_kem_set_params` set the NID specific parameters for a given PKEY. Why is this required? We use a higher level NID, namely EVP_PKEY_KEM for KEMS and EVP_PKEY_NISTDSA for PQ DSA. The PKEY has this NID, but doesn't know which specific algorithm is being referred to, so the `set_params` function sets algorithm specific parameters and methods.

For signature algorithms, we use a flurry of APIs that generate PKEYs based on OIDs, that will then require the specific algorithm OID to fully complete the PKEY structure. For example, consider the ASN.1 utility function:

`EVP_PKEY *EVP_parse_public_key(CBS *cbs);`

This function returns a newly-allocated `EVP PKEY` on success, however, this PKEY OID will be set to `EVP_PKEY_NISTDSA` and not a specific OID like `NID_MLDSA65`. Thus, to populate the PKEY with the correct parameters and methods, we call `EVP_PKEY_pqdsa_set_params` on the PKEY. You can see examples of this in the test code, for example, within `TEST_P(MLDSAParameterTest, MarshalParse) `:

```
// decode the DER structure, then parse as a PKEY.
  CBS cbs;
  CBS_init(&cbs, der, der_len);
  bssl::UniquePtr<EVP_PKEY> pkey_from_der(EVP_parse_public_key(&cbs));
  ASSERT_TRUE(pkey_from_der);

  // set the correct params for the PKEY
  EVP_PKEY_pqdsa_set_params(pkey_from_der.get(), nid);
  ```

Reply:
I've been working on an implementation more similar to how this is done for KEMs based upon Nevine's comments above.

Due to the single EVP PKEY type `EVP_PKEY_PQDSA` that covers different signature algorithms, we have the same issue as `EC` and `KEM` EVP PKEY types, in that a more specific algorithm identifier is required to set the appropriate sized/algorithm specific key buffers.

Rather than to use `EVP_PKEY_pqdsa_set_params` to retroactively set the correct params once a `PKEY` has been created, I have implemented `EVP_PKEY_pqdsa_new_raw_public_key` and `EVP_PKEY_pqdsa_new_raw_secret_key` that take a `nid` so they can assign the correct algorithm specific methods/params:

```
// EVP_PKEY_pqdsa_new_raw_public_key generates a new EVP_PKEY object of type
// EVP_PKEY_PQDSA, initializes the PQDSA key based on |nid| and populates the
// public key part of the PQDSA key with the contents of |in|. It returns the
// pointer to the allocated PKEY on sucess and NULL on error.
OPENSSL_EXPORT EVP_PKEY *EVP_PKEY_pqdsa_new_raw_public_key(int nid, const uint8_t *in, size_t len);

// EVP_PKEY_pqdsa_new_raw_secret_key generates a new EVP_PKEY object of type
// EVP_PKEY_PQDSA, initializes the PQDSA key based on |nid| and populates the
// secret key part of the PQDSA key with the contents of |in|. It returns the
// pointer to the allocated PKEY on sucess and NULL on error.
OPENSSL_EXPORT EVP_PKEY *EVP_PKEY_pqdsa_new_raw_secret_key(int nid, const uint8_t *in, size_t len);
```

[c0d4e65](https://github.com/aws/aws-lc/pull/1963/commits/c0d4e651b5accb0d1a05fc260635ce1d8968b3e8)